trajopt_sco/src/optimizers.cpp

```diff
@@ -271,11 +222,237 @@ struct MultiCritFilter {
+                "CONSTRAINTS",
+                "----------",
+                "----------",
+                "----------",
+                "---------"
+                "-");
+    for (size_t i = 0; i < old_cnt_viols.size(); ++i)
+    {
+      double approx_improve = old_cnt_viols[i] - model_cnt_viols[i];
+      double exact_improve = old_cnt_viols[i] - new_cnt_viols[i];
+      if (fabs(approx_improve) > 1e-8)
+        std::printf("%15s | %10.3e | %10.3e | %10.3e | %10.3e\n",
+                    cnt_names[i].c_str(),
+                    merit_error_coeff * old_cnt_viols[i],
+                    merit_error_coeff * approx_improve,
+                    merit_error_coeff * exact_improve,
+                    exact_improve / approx_improve);
+      else
+        std::printf("%15s | %10.3e | %10.3e | %10.3e | %10s\n",
+                    cnt_names[i].c_str(),
+                    merit_error_coeff * old_cnt_viols[i],
+                    merit_error_coeff * approx_improve,
+                    merit_error_coeff * exact_improve,
+                    "  ------  ");
+    }
+  }
+
+  std::printf("%15s | %10.3e | %10.3e | %10.3e | %10.3e\n",
+              "TOTAL",
+              old_merit,
+              approx_merit_improve,
+              exact_merit_improve,
+              merit_improve_ratio);
+}
+
+void BasicTrustRegionSQPResults::writeSolver(std::FILE* stream, bool header) const
+{
+  if (header)
+    std::fprintf(stream, "%s,%s,%s,%s,%s\n", "DESCRIPTION", "oldexact", "dapprox", "dexact", "ratio");
+
+  std::fprintf(stream,
+               "%s,%10.3e,%10.3e,%10.3e,%10.3e\n",
+               "Solver",
+               old_merit,
+               approx_merit_improve,
+               exact_merit_improve,
+               merit_improve_ratio);
+}
+
+void BasicTrustRegionSQPResults::writeVars(std::FILE* stream, bool header) const
+{
+  if (header)
+  {
+    std::fprintf(stream, "%s", "NAMES");
+    for (const auto& var : var_names)
+      std::fprintf(stream, ",%s", var.c_str());
+
+    std::fprintf(stream, "\n");
+  }
+
+  std::fprintf(stream, "%s", "VALUES");
+  for (const auto& x : new_x)
+    std::fprintf(stream, ",%e", x);
+
+  std::fprintf(stream, "\n");
+}
+
+void BasicTrustRegionSQPResults::writeCosts(std::FILE* stream, bool header) const
+{
+  if (header)
+  {
+    std::fprintf(stream, "%s", "COST NAMES");
+    for (const auto& name : cost_names)
+      std::fprintf(stream, ",%s", name.c_str());
+
+    std::fprintf(stream, "\nDESCRIPTION,%s,%s,%s,%s\n", "oldexact", "dapprox", "dexact", "ratio");
+  }
+
+  std::fprintf(stream, "%s", "COSTS");
+  for (size_t i = 0; i < old_cost_vals.size(); ++i)
+  {
+    double approx_improve = old_cost_vals[i] - model_cost_vals[i];
+    double exact_improve = old_cost_vals[i] - new_cost_vals[i];
+    std::fprintf(
+        stream, ",%e,%e,%e,%e", old_cost_vals[i], approx_improve, exact_improve, exact_improve / approx_improve);
+  }
+  std::fprintf(stream, "\n");
```

Comment:
It looks like we might want an fflush here. On my pc the data was not being written until the file was closed without it. That would mess up the live plotting code we had talked about.

Reply:
Done.